amethyst_ui/src/pass.rs

```diff
@@ -118,79 +148,39 @@ impl Pass for DrawUi {
 
     fn apply<'a, 'b: 'a>(
         &'a mut self,
-        supplier: Supplier<'a>,
-        (entities, screen_dimensions, mesh_storage, tex_storage, ui_image, ui_transform, ui_text): (
+        encoder: &mut Encoder,
+        effect: &mut Effect,
+        factory: Factory,
+        (
+            entities,
+            loader,
+            screen_dimensions,
+            mesh_storage,
+            tex_storage,
+            font_storage,
+            focused,
+            ui_image,
+            ui_transform,
+            mut ui_text,
+            editing,
+        ): (
             Entities<'a>,
+            Fetch<'a, Loader>,
             Fetch<'a, ScreenDimensions>,
             Fetch<'a, AssetStorage<Mesh>>,
             Fetch<'a, AssetStorage<Texture>>,
+            Fetch<'a, AssetStorage<FontAsset>>,
+            Fetch<'a, UiFocused>,
             ReadStorage<'a, UiImage>,
             ReadStorage<'a, UiTransform>,
-            ReadStorage<'a, UiText>,
+            WriteStorage<'a, UiText>,
+            ReadStorage<'a, TextEditing>,
         ),
-) -> DrawUiApply<'a>{
-        DrawUiApply {
-            entities,
-            screen_dimensions,
-            mesh_storage,
-            tex_storage,
-            ui_image,
-            ui_transform,
-            ui_text,
-            unit_mesh: self.mesh_handle.clone(),
-            cached_draw_order: &mut self.cached_draw_order,
-            supplier,
-        }
-    }
-}
-
-pub struct DrawUiApply<'a> {
-    entities: Entities<'a>,
-    screen_dimensions: Fetch<'a, ScreenDimensions>,
-    mesh_storage: Fetch<'a, AssetStorage<Mesh>>,
-    tex_storage: Fetch<'a, AssetStorage<Texture>>,
-    ui_image: ReadStorage<'a, UiImage>,
-    ui_transform: ReadStorage<'a, UiTransform>,
-    ui_text: ReadStorage<'a, UiText>,
-    unit_mesh: MeshHandle,
-    cached_draw_order: &'a mut CachedDrawOrder,
-    supplier: Supplier<'a>,
-}
-
-impl<'a> ParallelIterator for DrawUiApply<'a> {
-    type Item = ();
-
-    fn drive_unindexed<C>(self, consumer: C) -> C::Result
-    where
-        C: UnindexedConsumer<Self::Item>,
-    {
-        let DrawUiApply {
-            entities,
-            screen_dimensions,
-            mesh_storage,
-            tex_storage,
-            ui_image,
-            ui_transform,
-            ui_text,
-            unit_mesh,
-            cached_draw_order,
-            supplier,
-            ..
-        } = self;
-
-        let entities = &*entities;
-        let screen_dimensions = &screen_dimensions;
-        let mesh_storage = &mesh_storage;
-        let tex_storage = &tex_storage;
-        let ui_image = &ui_image;
-        let ui_text = &ui_text;
-        let ui_transform = &ui_transform;
-        let unit_mesh = &unit_mesh;
-
+    ) {
         // Populate and update the draw order cache.
         {
```

Comment:
yeah I thought about that, it turns out the overhead from those storage types actually ends up being worse than this setup, but only because this setup is very specifically optimized for tracking changes to one field.